akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -149,6 +133,98 @@ private[http] object OutgoingConnectionBlueprint {
 
   import ParserOutput._
 
+  private final class ResponsePrep(parserSettings: ParserSettings, eagerClose: Boolean)
+    extends GraphStage[FlowShape[ResponseOutput, HttpResponse]] {
+
+    private val in = Inlet[ResponseOutput]("ResponsePrep.in")
+    private val out = Outlet[HttpResponse]("ResponsePrep.out")
+
+    val shape = new FlowShape(in, out)
+
+    override def createLogic(effectiveAttributes: Attributes) = new GraphStageLogic(shape) {
+      private val idle = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case ResponseStart(statusCode, protocol, headers, entityCreator, _) ⇒
+            val entity = createEntity(entityCreator) withSizeLimit parserSettings.maxContentLength
+            push(out, HttpResponse(statusCode, headers, entity, protocol))
+
+          case MessageStartError(_, info) ⇒
+            throw IllegalResponseException(info)
+
+          case other ⇒
+            throw new IllegalStateException(f"ResponseStart expected but $other received.")
+        }
+      }
+
+      private val waitForMessageEnd = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case MessageEnd ⇒ setHandler(in, idle)
+          case other      ⇒ throw new IllegalStateException(f"MessageEnd expected but $other received.")
+        }
+      }
+
+      setHandler(in, idle)
+
+      private def createEntity(creator: EntityCreator[ResponseOutput, ResponseEntity]): ResponseEntity = {
+        creator match {
+          case StrictEntityCreator(entity) ⇒
+            pull(in)
+            setHandler(in, waitForMessageEnd)
```

Comment:
This style is fine IMO. We don't have a specific method for the scenario given (you may want to look at `read`, but it's not really better than just setting a handler like already done here).